rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+
+1. Completely rip out:
+   1. `Stage`
+   2. `Schedule`
+   3. `SystemDescriptor`
+   4. `IntoExclusiveSystem`
+   5. `SystemContainer` and friends
+2. Build out a basic schedule abstraction:
+   1. Create a `ScheduleLabel` trait
+   2. Store multiple schedules in a keyed `SystemRegistry` resource
+   3. Support adding systems to the schedule
+      1. These should be initialized by definition: see [PR #2777](https://github.com/bevyengine/bevy/pull/2777)
+   4. Support adding systems to specific schedules and create startup system API
+3. Rebuild core command processing machinery
+   1. Begin with trivial event-style commands
+   2. Explore how to regain parallelism
+4. Add system configuration tools
+   1. Create `SystemConfig` type
+   2. System sets store a `SystemConfig`
+   3. Allow systems to belong to sets
+   4. Store `ConfiguredSystems` in the schedule
+   5. Add `.add_systems` method, the `SystemGroup` type and the convenience methods for system groups
+   6. Special-case `CoreSet::First` and `CoreSet::Last` for convenience
+   7. Use [system builder syntax](https://github.com/bevyengine/rfcs/pull/31), rather than adding more complex methods to `App`
+5. Re-add strict ordering constraints
+6. Add run criteria
+   1. Create the machinery to generate run criteria from functions
+   2. Cache the accesses of run criteria on `ConfiguredSystem` in the schedule
+   3. Check the values of the run-criteria of systems before deciding whether it can be run
+7. Add states
+   1. Create `State` trait
+   2. Implement while-active states as sugar for simple run criteria
+   3. Create on-enter and on-exit schedules
+   4. Create sugar for adding systems to these schedules
+8. Rename "system chaining" to "system piping"
+   1. Usage was very confusing for new users
+   2. New concept of "systems with a linear graph of ordering constraints between them" is naturally described as a chain
+9. Add new examples
+    1. Complex control flow with supplementary schedules
+    2. Fixed time-step pattern
+10. Port the entire engine to the new paradigm
+    1. We almost certainly need to port the improved ambiguity checker over to make this new paradigm usable
+
+Given the massive scope, that sounds relatively straightforward!
+However, doing so will break approximately the entire engine, and tests will not pass again until step 10.
+
+Lets explore some of the more critical and challenging details.
+
+### Storing and configuring systems
+
+Schedules store systems in a configured, initialized form.
+
+```rust
+struct Schedule{
+   // Each schedule is associated with exactly one `World`, as systems must be initiliazed on it.
+
+   world_id: WorldId
+   // We need to be able to quickly look up specific systems and iterate over them.
+   // `SystemId` should be a unique opaque identifier, generated on system insertion.
+   // We cannot use a Vec + a `usize` index as the identifier,
+   // as it would not be stable when systems were removed, risking serious bugs.

```

Comment:
(creating a new thread for visibility because you resolved the last one) I agree with @inodentry and @hymm. Given how regularly we will be accessing data in this hashmap, this seems like it should be a Vec with generational indices. You _can_ support removals in that context (ex: with `Vec<Option<ConfiguredSystem>>` or by maintaining a dense `Vec<ConfiguredSystem>` and a dense mapping that maps id indices to actual indices).